src/expr/impl/src/scalar/encrypt.rs

```diff
@@ -0,0 +1,224 @@
+        })
+    }
+
+    // postgres example can be found [here](https://github.com/postgres/postgres/blob/master/contrib/pgcrypto/sql/rijndael.sql)
+    fn build_cipher(&self) -> Result<Cipher> {
+        // match config's algorithm, mode, padding to openssl's cipher
+        match (&self.algorithm, self.crypt_key.len(), &self.mode) {
+            (Algorithm::Blowfish, _, Mode::Cbc) => Ok(Cipher::bf_cbc()),
+            (Algorithm::Blowfish, _, Mode::Ecb) => Ok(Cipher::bf_ecb()),
+            (Algorithm::Aes, 16, Mode::Cbc) => Ok(Cipher::aes_128_cbc()),
+            (Algorithm::Aes, 16, Mode::Ecb) => Ok(Cipher::aes_128_ecb()),
+            (Algorithm::Aes, 24, Mode::Cbc) => Ok(Cipher::aes_192_cbc()),
+            (Algorithm::Aes, 24, Mode::Ecb) => Ok(Cipher::aes_192_ecb()),
+            (Algorithm::Aes, 32, Mode::Cbc) => Ok(Cipher::aes_256_cbc()),
+            (Algorithm::Aes, 32, Mode::Ecb) => Ok(Cipher::aes_256_ecb()),
+            _ => Err(ExprError::InvalidParam {
+                name: "mode",
+                reason: format!(
+                    "invalid algorithm {:?} mode: {:?}",
+                    self.algorithm, self.mode
+                )
+                .into(),
+            }),
+        }
+    }
+
+    fn enable_padding(&self) -> bool {
+        match self.padding {
+            Padding::Pkcs => true,
+            Padding::None => false,
+        }
+    }
+}
+
+/// from [pg doc](https://www.postgresql.org/docs/current/pgcrypto.html#PGCRYPTO-RAW-ENC-FUNCS)
+#[function(
+    "decrypt(bytea, bytea, varchar) -> bytea",
+    prebuild = "CipherConfig::parse_cipher_config($1, $2)?"
+)]
+pub fn decrypt(data: &[u8], config: &CipherConfig) -> Result<Box<[u8]>> {
+    let report_error = |e: ErrorStack| {
+        ExprError::Cryptography(Box::new(CryptographyError {
+            stage: CryptographyStage::Decrypt,
+            payload: data.into(),
+            reason: e.to_string().into(),
+        }))
+    };
+
+    let cipher = config.build_cipher()?;
+    let mut decrypter = Crypter::new(cipher, CipherMode::Decrypt, config.crypt_key.as_ref(), None)
+        .map_err(report_error)?;
+    decrypter.pad(config.enable_padding());
+    let mut decrypt = vec![0; data.len() + cipher.block_size()];
+    let count = decrypter.update(data, &mut decrypt).map_err(report_error)?;
+    let rest = decrypter
+        .finalize(&mut decrypt[count..])
+        .map_err(report_error)?;
+    decrypt.truncate(count + rest);
+    Ok(decrypt.into())
+}
+
+#[function(
+    "encrypt(bytea, bytea, varchar) -> bytea",
+    prebuild = "CipherConfig::parse_cipher_config($1, $2)?"
+)]
+pub fn encrypt(data: &[u8], config: &CipherConfig) -> Result<Box<[u8]>> {
+    let report_error = |e: ErrorStack| {
+        ExprError::Cryptography(Box::new(CryptographyError {
+            stage: CryptographyStage::Encrypt,
+            payload: data.into(),
+            reason: e.to_string().into(),
+        }))
+    };
+
+    let cipher = config.build_cipher()?;
```

Comment:
Does `prebuild = "CipherConfig::parse_cipher_config($1, $2)?"` produce `config: &CipherConfig` used in the function parameters

correct me if I am wrong, but `config` seems to be all the same when invoking multiple times, therefore `config.build_cipher()`'s returned results are all the same too, which is another thing that needs to be `prebuild`?

Reply:
Good catch. Looks like the `encryptor` can be prebuilt as well.

Reply:
I do agree `encryptor/decrypter` shall be encapsulated by `CipherConfig` as well. But they have internal states and operations take `&mut self`. `Clone` is not implemented either. So I guess we can only move `Crypter::new` and `cryptor.pad(config.enable_padding())` into the existing `build_{cipher,cryptor}` function, but they still need to be executed on every evaluation.

Reply:
> correct me if I am wrong, but config seems to be all the same when invoking multiple times, therefore config.build_cipher()'s returned results are all the same too, which is another thing that needs to be prebuild?

accoroding to [openssl doc](https://docs.rs/openssl/latest/openssl/symm/struct.Crypter.html#method.update), the `update` here takes a mut ref.

Reply:
resolved by #14869